src/librustc_mir/transform/inline.rs

```diff
@@ -94,33 +95,39 @@ impl Inliner<'tcx> {
 
         loop {
             local_change = false;
-            while let Some(callsite) = callsites.pop_front() {
+            'callsites: while let Some(callsite) = callsites.pop_front() {
                 debug!("checking whether to inline callsite {:?}", callsite);
                 if !self.tcx.is_mir_available(callsite.callee) {
                     debug!("checking whether to inline callsite {:?} - MIR unavailable", callsite);
                     continue;
                 }
 
-                let self_node_id = self.tcx.hir().as_local_node_id(self.source.def_id()).unwrap();
-                let callee_node_id = self.tcx.hir().as_local_node_id(callsite.callee);
-
-                let callee_body = if let Some(callee_node_id) = callee_node_id {
+                // This may result in a cycle since the callee MIR is from this crate
+                // and we be optimizing it right now in this very same query stack.
+                if callsite.callee.is_local() {
                     // Avoid a cycle here by only using `optimized_mir` only if we have
-                    // a lower node id than the callee. This ensures that the callee will
-                    // not inline us. This trick only works without incremental compilation.
-                    // So don't do it if that is enabled.
-                    if !self.tcx.dep_graph.is_fully_enabled()
-                        && self_node_id.as_u32() < callee_node_id.as_u32()
```

Comment:
I think you should keep this trick, 1) as a fast path to avoid doing the cycle check, 2) as a way to allow inlining in mutual recursion. It should be changed to work with incremental compilation by comparing `DefPath`s instead.

Reply:
It's probably cheaper to compare `def_path_hash`es instead. This would make it comparable to my suggestion but with less caching.

Reply:
This trick eliminates a lot of actually possible inlinings, so it has a very negative effect on runtime of programs when only MIR inlining is used (cg_clif)

Reply:
I think @Zoxc is talking about keeping the trick as a way to avoid checking for recursion, but *also* allowing inlining (with the recursion check) otherwise.